pilot/pkg/proxy/envoy/externalservice.go

```diff
@@ -0,0 +1,189 @@
+				listeners = append(listeners,
+					buildTCPListener(routeConfig, WildcardAddress, int(port.Number), modelPort.Protocol))
+			default:
+				// handled elsewhere
+			}
+		}
+	}
+
+	return listeners, clusters
+}
+
+func buildExternalServiceCluster(mesh *meshconfig.MeshConfig,
+	address, endpointPortName string, port *model.Port, labels model.Labels,
+	discovery routingv2.ExternalService_Discovery, endpoints []*routingv2.ExternalService_Endpoint) *Cluster {
+
+	service := model.Service{Hostname: address}
+	key := service.Key(port, labels)
+	clusterName := truncateClusterName(OutboundClusterPrefix + key)
+
+	// will only be populated with discovery type dns or static
+	hosts := make([]Host, 0)
+	for _, endpoint := range endpoints {
+		if !labels.SubsetOf(model.Labels(endpoint.Labels)) {
+			continue
+		}
+
+		var found bool
+		for name, portNumber := range endpoint.Ports {
+			if name == endpointPortName {
+				url := fmt.Sprintf("tcp://%s:%d", endpoint.Address, int(portNumber))
+				hosts = append(hosts, Host{URL: url})
+
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			// default to the external service port
+			url := fmt.Sprintf("tcp://%s:%d", endpoint.Address, int(port.Port))
+			hosts = append(hosts, Host{URL: url})
+		}
+	}
+
+	var clusterType, lbType string
+	switch discovery {
+	case routingv2.ExternalService_NONE:
+		clusterType = ClusterTypeOriginalDST
+		lbType = LbTypeOriginalDST
+	case routingv2.ExternalService_DNS:
+		clusterType = ClusterTypeStrictDNS
+		lbType = LbTypeRoundRobin
+	case routingv2.ExternalService_STATIC:
+		clusterType = ClusterTypeStatic
+		lbType = LbTypeRoundRobin
+	}
+
+	var sslContext interface{}
+	if port.Protocol == model.ProtocolHTTPS {
+		sslContext = &SSLContextExternal{}
+	}
+
+	var features string
+	switch port.Protocol {
+	case model.ProtocolHTTP2, model.ProtocolGRPC:
+		features = ClusterFeatureHTTP2
+	}
+
+	return &Cluster{
+		Name:             clusterName,
+		ServiceName:      key,
+		ConnectTimeoutMs: protoDurationToMS(mesh.ConnectTimeout),
+		Type:             clusterType,
+		LbType:           lbType,
+		Hosts:            hosts,
+		SSLContext:       sslContext,
+		Features:         features,
+		outbound:         true,
+		hostname:         address,
+		port:             port,
+		labels:           labels,
+	}
+}
+
+func buildExternalServiceVirtualHost(externalService *routingv2.ExternalService, portName, destination string,
+	mesh *meshconfig.MeshConfig, sidecar model.Node, port *model.Port, instances []*model.ServiceInstance,
+	config model.IstioConfigStore) *VirtualHost {
+
+	service := &model.Service{Hostname: destination}
+	buildClusterFunc := func(hostname string, port *model.Port, labels model.Labels, isExternal bool) *Cluster {
+		return buildExternalServiceCluster(mesh, hostname, portName, port, labels,
+			externalService.Discovery, externalService.Endpoints)
+	}
+
+	// FIXME: clusters generated if the routing rule routes traffic to other services will be constructed incorrectly
```

Comment:
What exactly do you mean here: that I can't have a route rule that sends 50% of traffic to A _and_ 50% to (some external service) B?

What does "constructed incorrectly" mean for the user? Envoy will reject the config, or it'll crash, or it'll fail to route at runtime, etc? More generally, is it a "oops, Pilot won't like it" or a "oops we broke production"?